aws_advanced_python_wrapper/read_write_splitting_plugin.py

```diff
@@ -147,209 +156,526 @@ def _update_internal_connection_info(self):
+            return False
 
     @staticmethod
-    def _get_writer(hosts: Tuple[HostInfo, ...]) -> Optional[HostInfo]:
-        for host in hosts:
+    def close_connection(conn: Optional[Connection], driver_dialect: DriverDialect):
+        if conn is not None:
+            try:
+                driver_dialect.execute(ReadWriteSplittingConnectionManager._CLOSE_METHOD, lambda: conn.close())
+            except Exception:
+                # Swallow exception
+                return
+
+
+class ConnectionHandler(Protocol):
+    """Protocol for handling writer/reader connection logic."""
+
+    @property
+    def host_list_provider_service(self) -> Optional[HostListProviderService]:
+        """Getter for the 'host_list_provider_service' attribute."""
+        ...
+
+    @host_list_provider_service.setter
+    def host_list_provider_service(self, new_value: int) -> None:
+        """The setter for the 'host_list_provider_service' attribute."""
+        ...
+
+    def open_new_writer_connection(
+        self,
+        plugin_service_connect_func: Callable[[HostInfo], Connection],
+    ) -> tuple[Optional[Connection], Optional[HostInfo]]:
+        """Open a writer connection."""
+        ...
+
+    def open_new_reader_connection(
+        self,
+        plugin_service_connect_func: Callable[[HostInfo], Connection],
+    ) -> tuple[Optional[Connection], Optional[HostInfo]]:
+        """Open a reader connection."""
+        ...
+
+    def get_verified_initial_connection(
+        self,
+        host_info: HostInfo,
+        is_initial_connection: bool,
+        plugin_service_connect_func: Callable[[HostInfo], Connection],
+        connect_func: Callable,
+    ) -> Connection:
+        """Verify initial connection or return normal workflow."""
+        ...
+
+    def should_update_writer_with_current_conn(
+        self, current_conn: Connection, current_host: HostInfo, writer_conn: Connection
+    ) -> bool:
+        """Return true if the current connection fits the criteria of a writer connection."""
+        ...
+
+    def should_update_reader_with_current_conn(
+        self, current_conn: Connection, current_host: HostInfo, reader_conn: Connection
+    ) -> bool:
+        """Return true if the current connection fits the criteria of a reader connection."""
+        ...
+
+    def is_writer_host(self, current_host: HostInfo) -> bool:
+        """Return true if the current host fits the criteria of a writer host."""
+        ...
+
+    def is_reader_host(self, current_host: HostInfo) -> bool:
+        """Return true if the current host fits the criteria of a writer host."""
+        ...
+
+    def old_reader_can_be_used(self, reader_host_info: HostInfo) -> bool:
+        """Return true if the current host can be used to switch connection to."""
+        ...
+
+    def need_connect_to_writer(self) -> bool:
+        """Return true if switching to reader should instead connect to writer."""
```

Comment:
I think this would make things a bit clearer
```suggestion
    def has_no_readers(self) -> bool:
        """Return true if there are no readers in the host list"""
```